applications/coreutils/tac.cpp

```diff
@@ -0,0 +1,122 @@
+#include <libsystem/cmdline/CMDLine.h>
+#include <libsystem/io/Path.h>
+#include <libsystem/io/Stream.h>
+#include <libutils/String.h>
+#include <libutils/StringBuilder.h>
+#include <libutils/Vector.h>
+
+static bool before = false;
+static char *separator = nullptr;
+
+static const char *usages[] = {
+    "NAME...",
+    "[OPTION]... NAME...",
+    nullptr,
+};
+
+static CommandLineOption options[] = {
+    COMMANDLINE_OPT_HELP,
+    COMMANDLINE_OPT_STRING("separator", 's', separator,
+                           "Choose the separator to be used instead of \\n",
+                           COMMANDLINE_NO_CALLBACK),
+    COMMANDLINE_OPT_BOOL("before", 'b', before,
+                         "Attach the separator before instead of after the string",
+                         COMMANDLINE_NO_CALLBACK),
+    COMMANDLINE_OPT_END};
+
+static CommandLine cmdline = CMDLINE(
+    usages,
+    options,
+    "Concatenate and print lines of file(s) in reverse.",
+    "If no filename provided read from input stream");
+
+Result tac(Stream *const input_stream)
+{
+    size_t read;
+    char buffer[1024];
+    char sep = (separator == nullptr ? '\n' : separator[0]);
+
+    StringBuilder s, temp;
+    Vector<String> lines(20);
+
+    while ((read = stream_read(input_stream, buffer, 1024)) != 0)
+    {
+        s.append(String(buffer, read));
+    }
+
+    String file_contents = s.finalize();
+
+    for (size_t i = 0; i < file_contents.length(); ++i)
+    {
+        if (file_contents[i] == sep || i == file_contents.length() - 1)
+        {
+            if (!before)
+            {
+                temp.append(file_contents[i]);
+            }
+
+            lines.push_back(temp.finalize());
```

Comment:
There's maybe a more optimal way to do that :thinking: 
(between the beginning of the `for` loop to its end)